File: web/play/src/ui/history.ts
// The client history: what each client asked, and what it was told.
//
// Act III's last levels are judged on this list. A read must observe every
// write that was acknowledged before it started, and a watermark must not go
// backwards — so the player must be able to read the history the game judges,
// operation by operation, in the order each client asked.
//
// Every field is read defensively: a client the engine sends with no reads, or
// an engine that adds a field tomorrow, renders the same.

import type { ClientView, GameView, ProposalView, ReadView } from '../types';
import { h } from '../render/dom';

/** What the game knows about one client operation. */
export type OpStatus = 'pending' | 'ambiguous' | 'proposed' | 'acked' | 'read';

/** One row of the history. */
export interface OpRow {
  /** The client that asked. */
  client: number;
  /** Whether it is a write. */
  write: boolean;
  /** `write 2` or `read 5`. */
  name: string;
  /** The command the client wrote, or an empty string for a read. */
  value: string;
  /** The node the client asked. */
  node: number;
  /** The slot a write landed at, or the watermark a read observed. */
  slot: number | null;
  /** What the game can say about it. */
  status: OpStatus;
  /** One sentence for the player. */
  detail: string;
}

function array<T>(raw: unknown): T[] {
  return Array.isArray(raw) ? (raw as T[]) : [];
}

function slotOf(raw: unknown): number | null {
  return typeof raw === 'number' && Number.isFinite(raw) ? raw : null;
}

function writeRow(client: ClientView, proposal: ProposalView, alive: boolean): OpRow {
  const slot = slotOf(proposal.slot);
  const value = typeof proposal.value === 'string' ? proposal.value : '';
  const base = {
    client: client.id,
    write: true,
    name: `write ${proposal.seq}`,
    value,
    node: proposal.node,
    slot,
  };
  if (proposal.acked === true) {
    return {
      ...base,
      status: 'acked',
      detail: `Node ${proposal.node} applied it at slot ${slot ?? '?'} and acknowledged it.`,
    };
  }
  if (!alive) {
    return {
      ...base,
      status: 'ambiguous',
      detail:
        slot === null
          ? `Node ${proposal.node} is down. The write can still commit, so the result is ambiguous.`
          : `Node ${proposal.node} is down, and it holds the write at slot ${slot}. The write can still commit, so the result is ambiguous.`,
    };
  }
  if (slot !== null) {
    return {
      ...base,
      status: 'proposed',
      detail: `Node ${proposal.node} holds it at slot ${slot}. The client has no acknowledgement yet.`,
    };
  }
  return {
    ...base,
    status: 'pending',
    detail: `The client waits for node ${proposal.node} to give it a slot.`,
  };
}

function readRow(client: ClientView, read: ReadView, alive: boolean): OpRow {
  const index = slotOf(read.index);
  const base = {
    client: client.id,
    write: false,
    name: `read ${read.ctx}`,
    value: '',
    node: read.node,
    slot: index,
  };
  if (read.served === true) {
    return {
      ...base,
      status: 'read',
      detail: `Node ${read.node} served it at the watermark ${index === null ? 'unknown' : `slot ${index}`}.`,
    };
  }
  if (!alive) {
    return {
      ...base,
      status: 'ambiguous',
      detail: `Node ${read.node} is down. The read has no answer, so the result is ambiguous.`,
    };
  }
  return {
    ...base,
    status: 'pending',
    detail: `Node ${read.node} must prove that it is the leader now. Until it does, the read waits.`,
  };
}

/**
 * Every client operation this world has, writes first, in the order the client
 * asked.
 */
export function clientOps(view: GameView): OpRow[] {
  const world = view.world;
  const alive = new Map<number, boolean>();
  for (const node of array<{ id: number; alive: boolean }>(world.nodes)) {
    alive.set(node.id, node.alive !== false);
  }
  const rows: OpRow[] = [];
  for (const client of array<ClientView>(world.clients)) {
    for (const proposal of array<ProposalView>(client.proposals)) {
      rows.push(writeRow(client, proposal, alive.get(proposal.node) !== false));
    }
    for (const read of array<ReadView>(client.reads)) {
      rows.push(readRow(client, read, alive.get(read.node) !== false));
    }
  }
  return rows;
}

const STATUS_WORDS: Record<OpStatus, string> = {
  pending: 'pending',
  ambiguous: 'ambiguous',
  proposed: 'at a slot',
  acked: 'acknowledged',
  read: 'served',
};

/** The client history panel, or nothing when the level has no clients. */
export function renderHistory(view: GameView): HTMLElement | null {
  const clients = array<ClientView>(view.world.clients);
  if (clients.length === 0) return null;
  const rows = clientOps(view);
  return h(
    'section',
    { class: 'history' },
    h('h2', {}, 'The client history'),
    h(
      'p',
      { class: 'small' },
      'The game judges this list. A read must observe every write that the cluster acknowledged before the read started.',
    ),
    rows.length === 0
      ? h('p', { class: 'empty' }, 'The clients did nothing yet. Ask a node to write a command.')
      : h(
          'ol',
          { class: 'history-list' },
          ...rows.map((row) =>
            h(
              'li',
              { class: `history-op status-${row.status}` },
              h('span', { class: 'history-client' }, `client ${row.client}`),
              h('span', { class: 'history-name' }, row.value ? `${row.name} ${row.value}` : row.name),
              h('span', { class: 'history-status' }, STATUS_WORDS[row.status]),
              h('span', { class: 'history-detail' }, row.detail),
            ),
          ),
        ),
  );
}

File: crates/paros-play/src/world/history.rs
//! The client half of the world: what each client asked for, when it asked,
//! and the history the linearizability judge reads.
//!
//! The client is the only party that knows its own program order, so this is
//! recorded client-side and judged client-side.

use std::collections::BTreeSet;

use paros_core::{ClientId, ClientSeq, NodeId, Slot};

use crate::world::World;

/// One client's writes.
#[derive(Clone, Debug)]
pub(super) struct Proposal {
    pub(super) seq: ClientSeq,
    pub(super) value: String,
    pub(super) node: NodeId,
    pub(super) slot: Option<Slot>,
    pub(super) acked: bool,
    /// When the client issued it, on the history's own monotone counter.
    pub(super) issued: u64,
    /// When it was acknowledged, on the same counter.
    pub(super) acked_at: Option<u64>,
}

/// One client's reads.
#[derive(Clone, Debug)]
pub(super) struct PendingRead {
    pub(super) ctx: u64,
    pub(super) node: NodeId,
    /// The index the round captured, recorded here because
    /// `Proposer::read_rounds` exposes no accessor for it.
    pub(super) index: Option<Slot>,
    /// The beat sequence an ack must carry to count, for display only.
    pub(super) required_seq: u64,
    /// Who has acked a qualifying beat since the read opened — the world's own
    /// tally, for the prompt's summary. The *judgement* always comes from a
    /// clone of the real `Proposer`.
    pub(super) acks: BTreeSet<NodeId>,
    /// Whether this is a **leaderless** read: a Phase-1 quorum's vote
    /// watermarks rather than a leader's beat acks. The two are served through
    /// the same `ReadState`, and only the client knows which it asked for.
    pub(super) leaderless: bool,
    pub(super) served: bool,
    /// When the client asked, on the history's own monotone counter.
    pub(super) issued: u64,
    /// When it was answered, on the same counter.
    pub(super) served_at: Option<u64>,
}

/// One client-visible operation, as the linearizability judge reads it.
///
/// The client is the only party that knows its own program order, so this is
/// recorded **client-side**: what it asked, when it asked, when it was
/// answered, and the one number the answer carries — the slot a write landed
/// at, or the watermark a read observed.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct HistoryOp {
    /// Whose operation it is.
    pub client: u64,
    /// The node it was sent to.
    pub node: NodeId,
    /// True for a write, false for a read.
    pub write: bool,
    /// A write's slot once acknowledged, or a read's watermark once served.
    /// `None` on a read means the empty prefix.
    pub at: Option<Slot>,
    /// When it was issued, on the history's monotone counter.
    pub started: u64,
    /// When it completed, or `None` while it is still outstanding. An
    /// outstanding operation constrains nothing: it may still complete later.
    pub completed: Option<u64>,
}

/// One client.
#[derive(Clone, Debug)]
pub(super) struct Client {
    pub(super) id: ClientId,
    pub(super) next_seq: u64,
    pub(super) proposals: Vec<Proposal>,
    pub(super) reads: Vec<PendingRead>,
}

impl World {
    /// Whether every client write that was admitted has been applied by the
    /// node that admitted it.
    #[must_use]
    pub fn all_writes_acked(&self) -> bool {
        self.clients
            .iter()
            .flat_map(|c| c.proposals.iter())
            .all(|p| p.acked)
    }

    /// The clients this level gave the player, in id order.
    #[must_use]
    pub fn clients(&self) -> Vec<u64> {
        self.clients.iter().map(|client| client.id.0).collect()
    }

    /// Every distinct value a client asked for, in the order it was first
    /// asked.
    ///
    /// A goal that must name a value reads it here rather than writing it
    /// down. A level's own reference proposes what it likes, and a player may
    /// propose anything at all; a goal that compares an applied log against a
    /// literal is asking whether the player typed the same word as the
    /// reference, which is not the rule any level teaches.
    #[must_use]
    pub fn proposed_values(&self) -> Vec<String> {
        let mut asked: Vec<(u64, String)> = self
            .clients
            .iter()
            .flat_map(|client| client.proposals.iter())
            .map(|proposal| (proposal.issued, proposal.value.clone()))
            .collect();
        asked.sort_by_key(|(issued, _)| *issued);
        let mut values: Vec<String> = Vec::new();
        for (_, value) in asked {
            if !values.contains(&value) {
                values.push(value);
            }
        }
        values
    }

    /// Every client operation, in the order it was issued — the history the
    /// linearizability judge reads.
    #[must_use]
    pub fn history(&self) -> Vec<HistoryOp> {
        let mut ops: Vec<HistoryOp> = Vec::new();
        for client in &self.clients {
            for proposal in &client.proposals {
                ops.push(HistoryOp {
                    client: client.id.0,
                    node: proposal.node,
                    write: true,
                    at: proposal.slot.filter(|_| proposal.acked),
                    started: proposal.issued,
                    completed: proposal.acked_at,
                });
            }
            for read in &client.reads {
                ops.push(HistoryOp {
                    client: client.id.0,
                    node: read.node,
                    write: false,
                    at: read.index.filter(|_| read.served),
                    started: read.issued,
                    completed: read.served_at,
                });
            }
        }
        ops.sort_by_key(|op| (op.started, op.client, !op.write));
        ops
    }

    /// Judge the recorded history by the three conditions a totally ordered
    /// log needs — no search, because the log *is* the order:
    ///
    /// 1. a committed read observes every write acknowledged before it began;
    /// 2. watermarks never move backwards across non-overlapping reads;
    /// 3. a write issued after a committed read lands **above** that read's
    ///    watermark.
    ///
    /// Operations that never completed constrain nothing: a write whose ack
    /// never arrived may still be chosen later, and that is not a violation of
    /// anything.
    ///
    /// `Ok(())` is the history being linearizable so far.
    ///
    /// # Errors
    ///
    /// The sentence naming the condition that failed and the two operations
    /// that failed it.
    pub fn linearizable(&self) -> Result<(), String> {
        let ops = self.history();
        let reads: Vec<&HistoryOp> = ops
            .iter()
            .filter(|op| !op.write && op.completed.is_some())
            .collect();
        let writes: Vec<&HistoryOp> = ops
            .iter()
            .filter(|op| op.write && op.completed.is_some())
            .collect();
        for read in &reads {
            let began = read.started;
            for write in &writes {
                let Some(acked) = write.completed else {
                    continue;
                };
                if acked < began && write.at > read.at {
                    return Err(format!(
                        "client {}'s read at node {} observed {}. Client {}'s write at {} was \
                         already acknowledged before that read began. A read must not go behind \
                         a write that completed before the read began.",
                        read.client,
                        read.node.0,
                        at(read.at),
                        write.client,
                        at(write.at)
                    ));
                }
            }
        }
        for earlier in &reads {
            let Some(done) = earlier.completed else {
                continue;
            };
            for later in &reads {
                if later.started >= done && later.at < earlier.at {
                    return Err(format!(
                        "a read at node {} observed {}. A read at node {} had already observed \
                         {} before it. A watermark must not move backwards.",
                        later.node.0,
                        at(later.at),
                        earlier.node.0,
                        at(earlier.at)
                    ));
                }
            }
        }
        for read in &reads {
            let Some(done) = read.completed else { continue };
            for write in &writes {
                if write.started > done && write.at <= read.at {
                    return Err(format!(
                        "client {}'s write landed at {}. That is at or below {}, which a read at \
                         node {} had already observed. A write issued after a read must land \
                         above the watermark of that read.",
                        write.client,
                        at(write.at),
                        at(read.at),
                        read.node.0
                    ));
                }
            }
        }
        Ok(())
    }

    /// The highest slot a client write has been acknowledged at — the write a
    /// later linearizable read must not read behind.
    #[must_use]
    pub fn highest_acked_slot(&self) -> Option<Slot> {
        self.clients
            .iter()
            .flat_map(|client| client.proposals.iter())
            .filter(|proposal| proposal.acked)
            .filter_map(|proposal| proposal.slot)
            .max()
    }
}

/// "nothing", or "slot 3" — how the goals and the history judge name a
/// watermark.
fn at(slot: Option<Slot>) -> String {
    slot.map_or_else(
        || "the empty prefix".to_string(),
        |s| format!("slot {}", s.0),
    )
}
